[img-helpers] Add/Edit Image display utilities (#357)
oddbird/oddleventy: content/sample/images.md

sub: Using nunjucks and the image utility classes
---

{% import 'embed.macros.njk' as embed %}

{{ embed.figure(
  data=[{
    img: 'blog/2023/graph.jpg',
    alt: 'sketched graph on paper'
  }],
  class='media-pull',
  attrs={'style': '--rowspan: span 2'}
) }}

### `media-pull` will position an image next to the text on wide enough screens

The `media-pull` class can be used to pull an image to the left of the content
at the `xlarge` screen size. Using a style attribute, you can define how many
rows to span using the css custom property `--rowspan`. The first image example
uses the `embed.figure` macro.

{{ embed.img(
  src='blog/2023/graph.jpg',
  alt='A collection of button screenshots from one project',
  attrs={
    'class': 'media-pull',
    'style': '--rowspan: span 2'
  }
) }}

### `media-pull` with the `embed.img` macro

When using the `embed.img` macro in a markdown file, the image is placed
within a `<p>` tag. We have a CSS helper that targets this nested image with
`display: contents` in order to add layout classes which will work in [most
modern browsers](https://caniuse.com/css-has). When unsupported, the image will
appear centered within the content column.

> 

Author: @mirisuzanne I added this second example specifically for testing the embed.img macro since it doesn't work in Firefox yet.

If you want an image to float to a side of the text, you need to first wrap the
image and text in a div with a class of `contain`.

<div class="contain">
{{ embed.img(
  src='work/coachhub/browse-coach-sitemap.jpg',
  alt='site map for browsing coaches',
  attrs={
    'class': 'align-left'
  }
) }}

To align an image to the left of the text, add the class `align-left`.
Images will only float if the screen is wide enough.
</div>

<div class="contain">
{{ embed.img(
  src='work/coachhub/browse-coach-sitemap.jpg',
  alt='site map for browsing coaches',
  attrs={
    'class': 'align-right'
  }
) }}

To align an image to the right of the text, add the class `align-right`.
Images will only float if the screen is wide enough.
</div>

<div class="contain">
{{ embed.img(
  src='work/coachhub/browse-coach-sitemap.jpg',
  alt='site map for browsing coaches',
  attrs={
    'class': 'align-center'
  }
) }}

To center align an image add the class `align-center`. This is only applicable
if your image is smaller than the column width.

</div>

## Extending images outside the content column

To extend an image further outside of the content column's width, add one of

> the following classes: `extend-small`, `extend-large`, or `extend-full`.

Author: @mirisuzanne I added `extend-small` back in this documentation since we can add this to the embed.img macro now. Also it is useful for showing how extend-small's width is the same as the embed.figure macro by default so it only makes sense to use it on embed.img.

When using the `embed.img` macro in a markdown file, the image is placed
within a `<p>` tag. We have a CSS helper that targets this nested image with
`display: contents` in order to add these layout classes in [most modern

> browsers](https://caniuse.com/css-has).

Author: I added this issue to the trello story https://trello.com/c/PkgEjxR2/122-when-you-use-the-embedfigure-macro-it-adds-an-empty-p-p-before-and-after

  alt='Miriam presenting',
  sizes='card'
) }}

{{ embed.figure(
  data=[{
    img: 'blog/2023/graph.jpg',
    alt: 'sketched graph on paper'
  }],
  class='media-pull',
  attrs={'style': '--rowspan: span 2'}
) }}

## `media-pull` will position an image next to the text on wide enough screens

> The `embed.figure` macro is the only one that has access to the outer element in

Author: yes. the embed.image is always wrapped in a <p> though... so you don't have access to add the class to the <p> that it is wrapped in unfortunately. I should have clarified that this is available on embed.figure but not embed.img (I didn't test it works on other macros, only those two).

> The `embed.figure` macro is the only one that has access to the outer element in

Author: While I agree that the image in a `<p>` tag isn't ideal, I thought we were stuck with that for reasons.

> The `embed.figure` macro is the only one that has access to the outer element in

Reviewer: Well, we're maybe stuck with that in markdown specifically, unless we write our own transform (we probably could, but it's not super simple). But not all of our images are used in markdown - we sometimes use these macros in nunjucks files. So it seems to me like we should document the actual limitation here, and think of extending images as 'valid, but dangerous inside markdown until we find a fix'. 

Here's a fix that might work in modern browsers:

```css
p:has(> picture:only-child) { display: contents; }
```

> The `embed.figure` macro is the only one that has access to the outer element in

Author: <3 `:has`  

When will Firefox support this? I can't wait to use it more.

{{ embed.img(
  src='work/coachhub/browse-coach-sitemap.jpg',
  alt='site map for browsing coaches',
  attrs={
    'class': 'align-left'
  }
) }}

To align an image to the left of the text, add the class `align-left`.
Images will only float left or right if the screen is wide enough.
</div>

<div class="contain">
{{ embed.img(
  src='work/coachhub/browse-coach-sitemap.jpg',
  alt='site map for browsing coaches',
  attrs={
    'class': 'align-right'
  }
) }}

To align an image to the right of the text, add the class `align-right`.
Images will only float left or right if the screen is wide enough.
</div>

<div class="contain">
{{ embed.img(
  src='work/coachhub/browse-coach-sitemap.jpg',
  alt='site map for browsing coaches',
  attrs={
    'class': 'align-center'
  }
) }}

To center align an image add the class `align-center`. This is only necessary
if your image is smaller than the column width.

</div>

## Extending images outside the content column

To extend an image further outside of the content column's width, add one of
the following classes: `extend-large`, or `extend-full`. This is only available
with the `embed.figure` macro as the `embed.img` macro's class doesn't have

> access to the outer container. Also, the embed.figure macro is wider than the
> content column by default, as opposed to the `embed.img` macro.

Reviewer: ```suggestion
access to the outer container. Note that the `embed.figure` macro extends wider than the
content column by default, as opposed to the `embed.img` macro.
```

> ### `embed.figure` without any entend class for comparison

Reviewer: ```suggestion
### `embed.figure` without any extend class for comparison
```

Seems like this could also be called `extend-small` - and available to other media?

{{ embed.figure(
  data=[{
    img: 'herman/herman-color-palette.jpg',
    alt: ''
  }]
) }}

### `extend-large`

{{ embed.figure(
  data=[{
    img: 'herman/herman-color-palette.jpg',
    alt: ''
  }],
  class='extend-large'
) }}

### `extend-full`
{{ embed.figure(
  data=[{
    img: 'herman/herman-color-palette.jpg',
    alt: 'Herman color palettes'
  }],
  class='extend-full',
  caption='I am a caption'

) }}

## Add a shadow to make an image stand out from the background

You can add the class `img-shadow` to the `embed.figure` and `embed.img` to
display a shadow around the image. This doesn't work on figure if you have a

> caption as the shadow will be around that entire container.

Reviewer: Note that the shadows are not visible in dark mode?
